From f61c1c8b1f7f03adcbc44e0f23ac5fcca70be62f Mon Sep 17 00:00:00 2001
From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Wed, 17 Oct 2018 17:22:10 +0200
Subject: [PATCH 14/45] glupload: dmabuf-direct: report driver limitations to
 debug log

Report in the DEBUG log if the driver does not support importing a given
format with linear modifiers non-externally.

https://bugzilla.gnome.org/show_bug.cgi?id=783521
(cherry picked from commit 3c8ac7de80a51cf51da86d322308b24a70080116)
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 gst-libs/gst/gl/egl/gsteglimage.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/gst-libs/gst/gl/egl/gsteglimage.c b/gst-libs/gst/gl/egl/gsteglimage.c
index 476166351a..16a9136502 100644
--- a/gst-libs/gst/gl/egl/gsteglimage.c
+++ b/gst-libs/gst/gl/egl/gsteglimage.c
@@ -637,13 +637,19 @@ _gst_egl_image_check_dmabuf_direct (GstGLContext * context, int fourcc)
       break;
   }
   g_free (formats);
-  if (i == num_formats)
+  if (i == num_formats) {
+    GST_DEBUG ("driver does not support importing fourcc %" GST_FOURCC_FORMAT,
+        GST_FOURCC_ARGS (fourcc));
     return FALSE;
+  }
 
   ret = gst_eglQueryDmaBufModifiersEXT (egl_display, fourcc, 0, NULL, NULL,
       &num_modifiers);
-  if (!ret || num_modifiers == 0)
+  if (!ret || num_modifiers == 0) {
+    GST_DEBUG ("driver does not report modifiers for fourcc %"
+        GST_FOURCC_FORMAT, GST_FOURCC_ARGS (fourcc));
     return FALSE;
+  }
 
   modifiers = g_new (EGLuint64KHR, num_modifiers);
   external_only = g_new (EGLBoolean, num_modifiers);
@@ -658,12 +664,18 @@ _gst_egl_image_check_dmabuf_direct (GstGLContext * context, int fourcc)
 
   for (i = 0; i < num_modifiers; ++i) {
     if (modifiers[i] == DRM_FORMAT_MOD_LINEAR) {
+      if (external_only[i]) {
+        GST_DEBUG ("driver only supports external import of fourcc %"
+            GST_FOURCC_FORMAT, GST_FOURCC_ARGS (fourcc));
+      }
       ret = !external_only[i];
       g_free (modifiers);
       g_free (external_only);
       return ret;
     }
   }
+  GST_DEBUG ("driver only supports non-linear import of fourcc %"
+      GST_FOURCC_FORMAT, GST_FOURCC_ARGS (fourcc));
   g_free (modifiers);
   g_free (external_only);
   return FALSE;
-- 
2.11.0

